api/api_ethereum_test.go

```diff
@@ -2431,3 +2435,160 @@ func TestEthereumAPI_GetRawTransactionByBlockNumberAndIndex(t *testing.T) {
+			account3: {Balance: common.Big0, Code: hexutil.MustDecode(codeRevertHello)},
+		}}
+
+		// blockchain
+		dbm    = database.NewMemoryDBManager()
+		db     = state.NewDatabase(dbm)
+		block  = gspec.MustCommit(dbm)
+		header = block.Header()
+		chain  = &testChainContext{header: header}
+
+		// tx arguments
+		KLAY     = hexutil.Big(*big.NewInt(params.KLAY))
+		mKLAY    = hexutil.Big(*big.NewInt(params.KLAY / 1000))
+		gas1000  = hexutil.Uint64(1000)
+		gas40000 = hexutil.Uint64(40000)
+		baddata  = hexutil.Bytes(hexutil.MustDecode("0xdeadbeef"))
+	)
+
+	any := gomock.Any()
+	getStateAndHeader := func(...interface{}) (*state.StateDB, *types.Header, error) {
+		// Return a new state for each call because the state is modified by EstimateGas.
+		state, err := state.New(block.Root(), db, nil, nil)
+		return state, header, err
+	}
+	getEVM := func(_ context.Context, msg blockchain.Message, state *state.StateDB, header *types.Header, vmConfig vm.Config) (*vm.EVM, func() error, error) {
+		// Taken from node/cn/api_backend.go
+		vmError := func() error { return nil }
+		context := blockchain.NewEVMContext(msg, header, chain, nil)
+		return vm.NewEVM(context, state, chainConfig, &vmConfig), vmError, nil
+	}
+	mockBackend.EXPECT().ChainConfig().Return(chainConfig).AnyTimes()
+	mockBackend.EXPECT().RPCGasCap().Return(common.Big0).AnyTimes()
+	mockBackend.EXPECT().StateAndHeaderByNumber(any, any).DoAndReturn(getStateAndHeader).AnyTimes()
+	mockBackend.EXPECT().StateAndHeaderByNumberOrHash(any, any).DoAndReturn(getStateAndHeader).AnyTimes()
+	mockBackend.EXPECT().GetEVM(any, any, any, any, any).DoAndReturn(getEVM).AnyTimes()
+
+	testcases := []struct {
+		args      EthTransactionArgs
+		expectErr string
+		expectGas uint64
+	}{
+		{ // simple transfer
+			args: EthTransactionArgs{
+				From:  &account1,
+				To:    &account2,
+				Value: &KLAY,
+			},
+			expectGas: 21000,
+		},
+		{ // simple transfer with insufficient funds
+			args: EthTransactionArgs{
+				From:  &account2, // sender has 0 KLAY
+				To:    &account1,
+				Value: &KLAY, // transfer 1 KLAY
+			},
+			expectErr: "insufficient balance for transfer",
+		},
+		{ // simple transfer with insufficient funds with nonzero gasPrice
+			args: EthTransactionArgs{
+				From:     &account2, // sender has 0 KLAY
+				To:       &account1,
+				Value:    &KLAY, // transfer 1 KLAY
+				GasPrice: &mKLAY,
+			},
+			expectErr: "insufficient funds for transfer",
+		},
+		{ // simple transfer too high gasPrice
+			args: EthTransactionArgs{
+				From:     &account1, // sender has 2 KLAY
+				To:       &account2,
+				Value:    &KLAY,  // transfer 1 KLAY
+				GasPrice: &mKLAY, // allowance = (2 - 1) / 0.001 = 1000 gas
+			},
+			expectErr: "gas required exceeds allowance",
+		},
+		{ // empty create
+			args:      EthTransactionArgs{},
+			expectGas: 53000,
+		},
+		{ // ignore too small gasLimit
+			args: EthTransactionArgs{
+				Gas: &gas1000,
+			},
+			expectGas: 53000,
```

Comment:
Having numeric constants (1000, 21000, 40000, 53000) is more intuitive to me because their numeric relations matter here (e.g. 1000 < 21000, 40000 < 53000)

Reply:
Yes it can be. Considering the semantics, the exact value that represents a constant variable seemed to be more precise for me.